src/sdam/topology.ts

```diff
@@ -66,20 +66,95 @@ const stateTransition = makeStateMachine({
   [STATE_CLOSING]: [STATE_CLOSING, STATE_CLOSED]
 });
 
-const DEPRECATED_OPTIONS = new Set([
-  'autoReconnect',
-  'reconnectTries',
-  'reconnectInterval',
-  'bufferMaxEntries'
-]);
-
 const kCancelled = Symbol('cancelled');
 const kWaitQueue = Symbol('waitQueue');
 
-interface Topology {
+type ServerSelectionCallback = Callback<Server>;
+interface ServerSelectionRequest {
+  serverSelector: ServerSelector;
+  transaction: Transaction;
+  callback: ServerSelectionCallback;
+  timer?: NodeJS.Timeout;
+  [kCancelled]?: boolean;
+}
+
+export interface Topology {
   destroy(): void;
 }
 
+interface TopologyPrivate {
+  /** the id of this topology */
+  id: number;
+  /** passed in options */
+  options: TopologyOptions;
+  /** initial seedlist of servers to connect to */
+  seedlist: ServerAddress[];
+  /** initial state */
+  state: string;
+  /** the topology description */
+  description: TopologyDescription;
+  serverSelectionTimeoutMS: number;
+  heartbeatFrequencyMS: number;
+  minHeartbeatFrequencyMS: number;
+  /** allow users to override the cursor factory */
+  Cursor: typeof Cursor;
+  /** A map of server instances to normalized addresses */
+  servers: Map<string, Server>;
+  /** Server Session Pool */
+  sessionPool: ServerSessionPool;
+  /** Active client sessions */
+  sessions: Set<ClientSession>;
+  credentials?: MongoCredentials;
+  clusterTime?: ClusterTime;
+  /** timers created for the initial connect to a server */
+  connectionTimers: TimerQueue;
+
+  /** related to srv polling */
+  srvPoller?: SrvPoller;
+  detectTopologyDescriptionChange?: (event: TopologyDescriptionChangedEvent) => void;
+  handleSrvPolling?: (event: SrvPollingEvent) => void;
+}
+
+interface ServerAddress {
+  host: string;
+  port: number;
+  domain_socket?: string;
+}
+
+interface TopologyOptions extends ServerOptions {
+  host: string;
+  port?: number;
+  credentials?: MongoCredentials;
+  serverSelectionTimeoutMS: number;
+  heartbeatFrequencyMS: number;
+  minHeartbeatFrequencyMS: number;
+  replicaSet?: string;
+  cursorFactory: typeof Cursor;
+  srvHost?: string;
+  srvPoller?: SrvPoller;
+  logger?: Logger;
+  loggerLevel?: string;
+  directConnection: boolean;
+
+  metadata: ClientMetadata;
+}
+
+interface CursorOptions {
+  topology?: typeof Topology;
+  maxTimeMS?: number;
+  cursorFactory?: typeof Cursor;
+}
+
+interface ConnectOptions {
+  readPreference?: ReadPreference;
+}
+
+type ServerSelector = Function;
```

Comment:
Is there a more descriptive type for a selector?

Reply:
Once I turned your extra rules on I was forced to better type `server_selection.ts`, so now we have a better type for this :+1: